spec/weather_retriever_spec.rb

```diff
@@ -0,0 +1,149 @@
+  end
+
+  after { sql.execute 'DROP TABLE weather' }
+
+  describe '#run' do
+    it 'collects the city name' do
+      set_up_retriever.run
+
+      expect(presenter).to have_received(:city_name).once
+    end
+
+    it 'does not call the client if the database is up to date' do
+      retriever = set_up_retriever
+
+      add_weather_row
+      retriever.run
+
+      expect(open_weather_stub).not_to have_been_made
+    end
+
+    context 'calls the client' do
+      it 'if database is empty' do
+        set_up_retriever.run
+
+        expect(open_weather_stub).to have_been_made.once
+      end
+
+      it 'if database is out of date' do
+        yesterday = 1_617_896_069
+        retriever = set_up_retriever
+
+        add_weather_row(yesterday)
+        retriever.run
+
+        expect(open_weather_stub).to have_been_made.once
+      end
+    end
+
+    context 'calls the error_log' do
+      it 'if client receives an error' do
+        open_weather_stub { |stub| stub.to_return(status: 404) }
+
+        set_up_retriever.run
+
+        expect(presenter).to have_received(:log_error).once
+      end
+
+      it 'if weather information is not as expected' do
+        bad_data = { weather: [{ description: 'Irrelivent' }], main: {} }
+        open_weather_stub { |stub| stub.to_return(status: 200, body: bad_data.to_json) }
+
+        set_up_retriever.run
+
+        expect(presenter).to have_received(:log_error).once
+      end
+    end
+
+    context 'does not call error_log' do
+      it 'if client receives weather body' do
+        open_weather_stub { |stub| stub.to_return(status: 200, body: weather_info) }
+
+        set_up_retriever.run
+
+        expect(presenter).not_to have_received(:log_error)
+      end
+
+      it 'if weather information is ok' do
+        open_weather_stub { |stub| stub.to_return(status: 200, body: weather_info) }
+
+        set_up_retriever.run
+
+        expect(presenter).not_to have_received(:log_error)
+      end
+    end
+
+    it 'calls a presenter method with up to date weather' do
+      open_weather_stub { |stub| stub.to_return(status: 200, body: weather_info) }
+
+      set_up_retriever.run
+
+      expect(presenter).to have_received(:collect_weather).with(todays_weather)
+    end
+  end
+
+  def set_up_retriever
+    described_class.new(
+      OpenWeatherClient.new(CityNameConverter.new('fixtures/london_city_id.yaml')),
+      SQLDatabase.new('file::memory:?cache=shared'),
+      presenter
+    )
+  end
+
+  def presenter_stub
+    allow(presenter).to receive(:city_name) { 'London' }
+    allow(presenter).to receive(:collect_weather).with(kind_of(Hash))
+    allow(presenter).to receive(:log_error).with(kind_of(String))
+  end
+
+  def open_weather_stub
+    url = "#{ENV['API_URL']}/weather"
+    stub = stub_request(:get, url).with(
+      query: {
+        'id' => '2643743',
+        'units' => 'metric',
+        'appid' => ENV['API_KEY']
+      }
+    )
+    yield stub if block_given?
+    stub
+  end
```

Comment:
I think I would simplify this into a simple method with named arguments and default values:

```ruby
  def open_weather_stub(status: 200, body: weather_info)
    url = "#{ENV['API_URL']}/weather"
    query = {
      'id' => '2643743',
      'units' => 'metric',
      'appid' => ENV['API_KEY']
    }
    stub_request(:get, url).with(query: query).to_return(status: status, body: body)
  end
```

and then you just have to call it very simply as:

```ruby
open_weather_stub
```

or if you need to get a hold of the stub:

```ruby
stub = open_weather_stub
```

or if you want to pass a different value:

```ruby
open_weather_stub(body: bad_data.to_json)
open_weather_stub(status: 404)
open_weather_stub(status: 301, body: 'Redirecting...')
```